controller/api/destination/endpoint_profile_translator.go

```diff
@@ -1,69 +1,136 @@
+//
+// We omit ip and port labels because they are high cardinality.
+var endpointProfileUpdatesQueueOverflowCounter = promauto.NewCounter(
+	prometheus.CounterOpts{
+		Name: "endpoint_profile_updates_queue_overflow",
+		Help: "A counter incremented whenever the endpoint profile updates queue overflows",
+	},
+)
+
+// newEndpointProfileTranslator translates pod updates and profile updates to
 // DestinationProfiles for endpoints
 func newEndpointProfileTranslator(
 	enableH2Upgrade bool,
 	controllerNS,
 	identityTrustDomain string,
 	defaultOpaquePorts map[uint32]struct{},
-	log *logging.Entry,
 	stream pb.Destination_GetProfileServer,
+	endStream chan struct{},
+	log *logging.Entry,
 ) *endpointProfileTranslator {
 	return &endpointProfileTranslator{
 		enableH2Upgrade:     enableH2Upgrade,
 		controllerNS:        controllerNS,
 		identityTrustDomain: identityTrustDomain,
 		defaultOpaquePorts:  defaultOpaquePorts,
-		stream:              stream,
-		log:                 log.WithField("component", "endpoint-profile-translator"),
+
+		stream:    stream,
+		endStream: endStream,
+		updates:   make(chan *watcher.Address, updateQueueCapacity),
+		stop:      make(chan struct{}),
+
+		log: log.WithField("component", "endpoint-profile-translator"),
 	}
 }
 
+// Start initiates a goroutine which processes update events off of the
+// endpointProfileTranslator's internal queue and sends to the grpc stream as
+// appropriate. The goroutine calls non-thread-safe Send, therefore Start must
+// not be called more than once.
+func (ept *endpointProfileTranslator) Start() {
+	go func() {
+		for {
+			select {
+			case update := <-ept.updates:
+				ept.update(update)
+			case <-ept.stop:
+				return
+			}
+		}
+	}()
+}
+
+// Stop terminates the goroutine started by Start.
+func (ept *endpointProfileTranslator) Stop() {
+	close(ept.stop)
+}
+
 // Update sends a DestinationProfile message into the stream, if the same
 // message hasn't been sent already. If it has, false is returned.
 func (ept *endpointProfileTranslator) Update(address *watcher.Address) (bool, error) {
+	select {
+	case ept.updates <- address:
+		// Update has been successfully enqueued.
+	default:
+		// We are unable to enqueue because the channel does not have capacity.
+		// The stream has fallen too far behind and should be closed.
+		endpointProfileUpdatesQueueOverflowCounter.Inc()
+		select {
+		case <-ept.endStream:
+			// The endStream channel has already been closed so no action is
+			// necessary.
+		default:
+			ept.log.Error("Profile update queue full; aborting stream")
+			close(ept.endStream)
+		}
+	}
+	return true, nil
+}
+
+func (ept *endpointProfileTranslator) update(address *watcher.Address) {
 	opaquePorts := watcher.GetAnnotatedOpaquePorts(address.Pod, ept.defaultOpaquePorts)
 	endpoint, err := ept.createEndpoint(*address, opaquePorts)
 	if err != nil {
-		return false, fmt.Errorf("failed to create endpoint: %w", err)
+		ept.log.Error(err)
```

Comment:
Before this change no message was sent back when an error occurred; now we're sending a profile with an empty Endpoint. Is that intended?